### PR TITLE
MES-6240 Load journal data from MS

### PR DESCRIPTION
**Pull request by tomcrawleyy**
## Description
Introduces the functionality required for the journal provider to interact and pull journal data from the micro-service & to integrate mock data when serving locally. Migrates the slot provider in order to format + structure the journal.

## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [x] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -13,6 +13,7 @@
   ],
   "rules": {
     "@typescript-eslint/lines-between-class-members": "off",
+    "@typescript-eslint/no-unused-vars": "warn",
```
Comment (szabikr) [inline .eslintrc]: Dangerous move.. I would add a TODO when to remove this, and even then, I can see the potential of this going along with the development of the entire project.

File: src/app/providers/data-store/data-store.ts [hunk 2]
```diff
@@ -76,6 +76,7 @@ export class DataStoreProvider {
 
       // TODO: Send error through the logging service
 
+      // eslint-disable-next-line no-console
```
Comment (szabikr) [inline src/app/providers/data-store/data-store.ts]: Now that we have the logging mechanism in place, this error could be sent up with the other logs

File: src/store/journal/journal.actions.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
+// eslint-disable-next-line max-classes-per-file
```
Comment (szabikr) [inline src/store/journal/journal.actions.ts]: Are we not going to use the new syntax of NGRX which would cut the file size almost in half?

File: src/store/journal/journal.effects.ts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  private getRelevantSlotItemsByDate = (slotItems: SlotItem[]): { [date: string]: SlotItem[] } => {
+    let slotItemsByDate: { [date: string]: SlotItem[] };
+    slotItemsByDate = groupBy(slotItems as any, this.slotProvider.getSlotDate) as any;
+    slotItemsByDate = this.slotProvider.extendWithEmptyDays(slotItemsByDate);
+    slotItemsByDate = this.slotProvider.getRelevantSlots(slotItemsByDate);
+    return slotItemsByDate;
+  };
+
+  @Effect()
+  journal$ = this.actions$.pipe(
```
Comment (szabikr) [inline src/store/journal/journal.effects.ts]: `createEffect` could be used here (new NGRX syntax)

File: src/store/journal/journal.effects.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        startWith(null),
+      );
+      const pollTimer$ = manualRefreshes$.pipe(
+        switchMap(() => interval(this.appConfig.getAppConfig().journal.autoRefreshInterval)),
+      );
+
+      const pollsWhileOnline$ = pollTimer$
+        .pipe(
+          withLatestFrom(this.networkStateProvider.onNetworkChange()),
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (szabikr) [inline src/store/journal/journal.effects.ts]: can remove the `_` and reenable eslint

File: src/store/journal/journal.effects.ts [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+          select(getJournalState),
+          map(getSelectedDate),
+        ),
+        this.store$.pipe(
+          select(getJournalState),
+          map((journal) => canNavigateToPreviousDay(journal, this.dateTimeProvider.now())),
+        ),
+      ),
+    )),
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-shadow
```
Comment (szabikr) [inline src/store/journal/journal.effects.ts]: `[, , canNavigateToPreviousDay]` passes the lining

same down below

File: src/store/journal/journal.reducer.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  lastRefreshed: null,
+  slots: {},
+  selectedDate: '',
+  examiner: null,
+  completedTests: [],
+};
+
+/* eslint-disable */
+
+export function journalReducer(state = initialState, action: journalActions.JournalActionTypes): JournalModel {
```
Comment (szabikr) [inline src/store/journal/journal.reducer.ts]: `createReducer` function could be used to follow new NGRX syntax

**Review by szabikr — COMMENTED** [review]
Added a few comments, curious on you thoughts

File: .eslintrc (hunk 1 above)
Comment (ammarhaiderbjss) [reply .eslintrc]: Good point, this is removed now

File: src/store/journal/journal.effects.ts (hunk 6 above)
Comment (ammarhaiderbjss) [reply src/store/journal/journal.effects.ts]: done

File: src/store/journal/journal.effects.ts (hunk 4 above)
Comment (tomcrawleyy) [reply src/store/journal/journal.effects.ts]: done

File: mock/local-journal-non-test-activities.json [hunk 8]
```diff
@@ -7,7 +7,7 @@
     {
       "slotDetail": {
         "slotId": 1000,
-        "start": "2020-12-01T09:00:00",
+        "start": "2021-01-25T09:00:00",
```
Comment (szabikr) [inline mock/local-journal-non-test-activities.json]: Ohh you might wanna add these mock files to `.gitignore`

File: src/app/providers/slot/slot.actions.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { Action } from '@ngrx/store';
+
+export const SLOT_HAS_CHANGED = '[SlotProvider] SLOT_HAS_CHANGED';
+
+export class SlotHasChanged implements Action {
```
Comment (piercemorris) [inline src/app/providers/slot/slot.actions.ts:5]: These actions will also need to be updated but can do it at a later time (in my story perhaps)

**Review by piercemorris — APPROVED** [review]
Looks good!

File: mock/local-journal-non-test-activities.json [hunk 10]
```diff
@@ -7,7 +7,7 @@
     {
       "slotDetail": {
         "slotId": 1000,
-        "start": "2020-12-01T09:00:00",
+        "start": "2021-01-27T09:00:00",
```
Comment (szabikr) [inline mock/local-journal-non-test-activities.json]: I think the file might have to be manually untracked. A quick description is available in this article:
https://www.pluralsight.com/guides/how-to-use-gitignore-file
under the _What If I Already Have It Checked In?_ title

**Review by szabikr — APPROVED** [review]
alright yeah looks good

Just untrack the mock files

**Pull request merged**